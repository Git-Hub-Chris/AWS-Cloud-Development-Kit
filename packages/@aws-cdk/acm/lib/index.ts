export * from './certificate';
export * from './certificate-ref';

// The L1 library for AWS::CertificateManager:
export * from '../cfn/certificatemanager';
