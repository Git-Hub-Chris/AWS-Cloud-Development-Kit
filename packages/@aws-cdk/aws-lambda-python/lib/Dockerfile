# The correct AWS SAM build image based on the runtime of the function will be
# passed as build arg. The default allows to do `docker build .` when testing.
ARG IMAGE=public.ecr.aws/sam/build-python3.7
FROM $IMAGE

ARG PIP_INDEX_URL
ARG PIP_EXTRA_INDEX_URL
ARG HTTPS_PROXY

# Upgrade pip (required by cryptography v3.4 and above, which is a dependency of poetry)
RUN pip install --upgrade pip

# pipenv 2022.4.8 is the last version with Python 3.6 support
# poetry>=1.2 cannot use MD5 hashes for installing packages, which is the only hashes
# AWS Code Artifact supports.
# https://github.com/python-poetry/poetry/issues/6301
RUN pip install pipenv==2022.4.8 "poetry<1.2"

CMD [ "python" ]
