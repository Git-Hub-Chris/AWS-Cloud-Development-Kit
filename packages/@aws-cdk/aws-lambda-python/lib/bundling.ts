import * as fs from 'fs';
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import * as cdk from '@aws-cdk/core';

/**
 * Dependency files to exclude from the asset hash.
 */
export const DEPENDENCY_EXCLUDES = ['*.pyc'];

/**
 * The location in the image that the bundler image caches dependencies.
 */
export const BUNDLER_DEPENDENCIES_CACHE = '/var/dependencies';

export interface BundleOptions {
  entry: string;
  runtime: lambda.Runtime;
  depsCommand: string;
}

export function bundle(options: BundleOptions) {
  const { entry, runtime, depsCommand } = options;

  // Determine which bundling image to use.
  let image: cdk.BundlingDockerImage;
  if (hasDependencies(entry)) {
    // When dependencies are present, we use a Dockerfile that can create a
    // cacheable layer. We can't use this Dockerfile if there aren't
    // dependencies or the Dockerfile will complain about missing sources.
    image = cdk.BundlingDockerImage.fromAsset(entry, {
      buildArgs: {
        IMAGE: runtime.bundlingDockerImage.image,
      },
      file: path.join(__dirname, 'Dockerfile.dependencies'),
    });
  } else {
    // When dependencies aren't present, we can use the default Dockerfile that
    // doesn't need to create a cacheable layer.
    image = cdk.BundlingDockerImage.fromAsset(entry, {
      buildArgs: {
        IMAGE: runtime.bundlingDockerImage.image,
      },
      file: path.join(__dirname, 'Dockerfile'),
    });
  }

  return lambda.Code.fromAsset(entry, {
    assetHashType: cdk.AssetHashType.BUNDLE,
    exclude: DEPENDENCY_EXCLUDES,
    bundling: {
      image,
      command: ['bash', '-c', depsCommand],
    },
  });
}

export interface BundleFunctionOptions {
  readonly entry: string;
  readonly runtime: lambda.Runtime;
}

export function bundleFunction(options: BundleFunctionOptions): lambda.AssetCode {
  const { entry, runtime } = options;

  const depsCommand = chain([
    hasDependencies(entry) ? `rsync -r ${BUNDLER_DEPENDENCIES_CACHE}/. ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}` : '',
    `rsync -r . ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}`,
  ]);

  return bundle({
    entry,
    runtime,
    depsCommand,
  });
}

export interface BundleDependencyLayerOptions {
  readonly entry: string;
  readonly runtime: lambda.Runtime;
}

export function bundleDependenciesLayer(options: BundleDependencyLayerOptions): lambda.AssetCode {
  const { entry, runtime } = options;

  const depsCommand = chain([
    `rsync -r ${BUNDLER_DEPENDENCIES_CACHE} ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}/python`,
  ]);

  return bundle({
    entry,
    runtime,
    depsCommand,
  });
}

export interface BundleFilesLayerOptions extends cdk.CopyOptions {
  readonly entry: string;
}

export function bundleFilesLayer(options: BundleFilesLayerOptions) {
  // Use local bundling instead of docker to copy the user's code into a
  // subdirectory while respecting the given excludes. We can't easily do this
  // in the container as the container doesn't have access to
  // `FileSystem.copyDirectory` that accepts the excludes.
  return lambda.Code.fromAsset(options.entry, {
    assetHashType: cdk.AssetHashType.BUNDLE,
    bundling: {
      // Local bundling is employed instead of running docker, but the bundler
      // still wants `image`. So, I've rigged the docker component of this
      // bundle to emit an error if the bundler runs docker.
      image: cdk.BundlingDockerImage.fromRegistry('alpine'),
      command: ['sh', '-c', 'echo The bundler attempted to run docker && exit 1'],
      local: new PythonCodeLocalBundler(options),
    },
  });
}

/**
 * A local bundling implementation which copies files to a python subdirectory
 * in the emitted asset.
 */
export class PythonCodeLocalBundler implements cdk.ILocalBundling {
  constructor(private options: BundleFilesLayerOptions) {}

  tryBundle(outputDir: string) {
    const { entry } = this.options;
    const layerSubDir = path.join(outputDir, 'python');
    try {
      fs.mkdirSync(layerSubDir);
      cdk.FileSystem.copyDirectory(entry, layerSubDir, {
        ...this.options,
      });
      return true;
    } catch (err) {
      throw new Error(`Failed to copy ${entry} to ${layerSubDir}: ${err.toString()}`);
    }
  }
}

/**
 * Checks to see if the `entry` directory contains a type of dependency that
 * we know how to install.
 */
export function hasDependencies(entry: string): boolean {
  if (fs.existsSync(path.join(entry, 'Pipfile'))) {
    return true;
  }

  if (fs.existsSync(path.join(entry, 'requirements.txt'))) {
    return true;
  }

  return false;
}

function chain(commands: string[]): string {
  return commands.filter(c => !!c).join(' && ');
}
