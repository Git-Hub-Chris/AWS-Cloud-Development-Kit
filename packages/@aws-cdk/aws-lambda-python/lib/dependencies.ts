import * as fs from 'fs';
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import * as cdk from '@aws-cdk/core';

export function hasDependencies(entry: string): boolean {
  return getDependenciesType(entry) != DependencyType.NONE;
}

export interface InstallDependenciesCommandsOptions {
  entry: string;
  runtime: lambda.Runtime;
  outputPath: string;
}

export function installDependenciesCommands(options: InstallDependenciesCommandsOptions): string[] {
  const entry = options.entry;

  const dependenciesType = getDependenciesType(entry);

  switch (dependenciesType) {
    case DependencyType.REQUIREMENTS:
      return pipInstallCommands(options);
    case DependencyType.PIPENV:
      return pipenvInstallCommands(options);
    case DependencyType.NONE:
      return [];
    default:
      throw new Error('Unsupported dependencies type');
  }
}

export enum DependencyType {
  NONE = 'none',
  REQUIREMENTS = 'requirements',
  PIPENV = 'pipenv',
}

export function getDependenciesType(entry: string): DependencyType {
  if (fs.existsSync(path.join(entry, 'Pipfile'))) {
    return DependencyType.PIPENV;
  }

  if (fs.existsSync(path.join(entry, 'requirements.txt'))) {
    return DependencyType.REQUIREMENTS;
  }

  return DependencyType.NONE;
}

interface PipInstallCommandsOptions extends InstallDependenciesCommandsOptions {
  requirementsTxtPath?: string;
}

export function pipInstallCommands(options: PipInstallCommandsOptions): string[] {
  const runtime = options.runtime;
  const requirementsTxtPath = options.requirementsTxtPath ?? 'requirements.txt';
  const outputPath = options.outputPath ?? cdk.AssetStaging.BUNDLING_OUTPUT_DIR;
  const pipCommand = getPipCommand(runtime);

  return [
    `${pipCommand} install -r ${requirementsTxtPath} -t ${outputPath}`,
  ];
}

export function getPipCommand(runtime: lambda.Runtime): string {
  return runtime === lambda.Runtime.PYTHON_2_7 ? PipCommand.PIP : PipCommand.PIP3;
}

enum PipCommand {
  /**
   * Used by Runtime.PYTHON_2_7
   */
  PIP = 'pip',

  /**
   * Used for Runtimes other than 2.7
   */
  PIP3 = 'pip3',
}

export function pipenvInstallCommands(options: InstallDependenciesCommandsOptions) {
  const pipenvCommand = 'HOME=/tmp/pipenv python -m pipenv';
  const requirementsTxtPath = '/tmp/requirements.txt';

  return [
    `${pipenvCommand} lock -r >${requirementsTxtPath}`,
    ...pipInstallCommands({ ...options, requirementsTxtPath }),
  ];
}
