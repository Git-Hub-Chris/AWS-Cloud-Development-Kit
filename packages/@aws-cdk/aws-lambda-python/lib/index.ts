export * from './function';
export * from './layer';
export * from './packaging';
