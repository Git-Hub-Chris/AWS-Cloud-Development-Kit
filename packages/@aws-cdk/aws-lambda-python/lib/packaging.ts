import * as fs from 'fs';
import * as path from 'path';

export enum DependenciesFile {
  PIP = 'requirements.txt',
  POETRY = 'poetry.lock',
  PIPENV = 'Pipfile.lock',
  NONE = ''
}

export interface PackagingProps {
  /**
   * Dependency file for the type of packaging.
   */
  readonly dependenciesFile: DependenciesFile;
  /**
   * Command to export the dependencies into a pip-compatible `requirements.txt` format.
   *
   * @default - No dependencies are exported.
   */
  readonly exportCommand?: string;
}

class PoetryPackagingProps {
  /**
   * Whether to export Poetry dependencies without hashes. This can fix build issues when some dependencies are exporting
   * with hashes and others are not, causing pip to fail the build.
   *
   * @see https://github.com/aws/aws-cdk/issues/19232
   * @default Hashes are included in the exported `requirements.txt` file
   */
  readonly poetryExcludeHashes?: boolean;
}

export class Packaging {

  /**
   * Standard packaging with `pip`.
   */
  public static withPip(): Packaging {
    return new Packaging({
      dependenciesFile: DependenciesFile.PIP,
    });
  }

  /**
   * Packaging with `pipenv`.
   */
  public static withPipenv(): Packaging {
    return new Packaging({
      dependenciesFile: DependenciesFile.PIPENV,
      // By default, pipenv creates a virtualenv in `/.local`, so we force it to create one in the package directory.
      // At the end, we remove the virtualenv to avoid creating a duplicate copy in the Lambda package.
      exportCommand: `PIPENV_VENV_IN_PROJECT=1 pipenv lock -r > ${DependenciesFile.PIP} && rm -rf .venv`,
    });
  }

  /**
   * Packaging with `poetry`.
   */
  public static withPoetry(props?: PoetryPackagingProps) {
    return new Packaging({
      dependenciesFile: DependenciesFile.POETRY,
      // Export dependencies with credentials available in the bundling image.
      exportCommand: props?.poetryExcludeHashes ?
        `poetry export --without-hashes --with-credentials --format ${DependenciesFile.PIP} --output ${DependenciesFile.PIP}` :
        `poetry export --with-credentials --format ${DependenciesFile.PIP} --output ${DependenciesFile.PIP}`,
    });
  }

  /**
   * No dependencies or packaging.
   */
  public static withNoPackaging(): Packaging {
    return new Packaging({ dependenciesFile: DependenciesFile.NONE });
  }

  public static fromEntry(entry: string, poetryExcludeHashes?: boolean): Packaging {
    if (fs.existsSync(path.join(entry, DependenciesFile.PIPENV))) {
      return this.withPipenv();
    } if (fs.existsSync(path.join(entry, DependenciesFile.POETRY))) {
      return this.withPoetry({ poetryExcludeHashes });
    } else if (fs.existsSync(path.join(entry, DependenciesFile.PIP))) {
      return this.withPip();
    } else {
      return this.withNoPackaging();
    }
  }

  public readonly dependenciesFile: string;
  public readonly exportCommand?: string;
  constructor(props: PackagingProps) {
    this.dependenciesFile = props.dependenciesFile;
    this.exportCommand = props.exportCommand;
  }
}