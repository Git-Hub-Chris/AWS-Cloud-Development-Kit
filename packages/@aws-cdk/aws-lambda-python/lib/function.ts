import * as fs from 'fs';
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import * as cdk from '@aws-cdk/core';
import { bundle } from './bundling';

/**
 * Properties for a PythonFunction
 */
export interface PythonFunctionProps extends lambda.FunctionOptions {
  /**
   * Path to the (Python) entry file.
   */
  readonly entry: string;

  /**
   * The name of the exported handler in the entry file.
   *
   * @default handler
   */
  readonly handler?: string;

  /**
   * The runtime environment. Only runtimes of the Python family are
   * supported.
   *
   * @default lambda.Runtime.PYTHON_3_7
   */
  readonly runtime?: lambda.Runtime;
}

/**
 * A Python Lambda function
 */
export class PythonFunction extends lambda.Function {
  constructor(scope: cdk.Construct, id: string, props: PythonFunctionProps) {
    if (props.runtime && props.runtime.family !== lambda.RuntimeFamily.PYTHON) {
      throw new Error('Only `PYTHON` runtimes are supported.');
    }

    // Entry and defaults
    const entry = resolveEntry(props.entry);
    const handler = props.handler ?? 'handler';
    const runtime = props.runtime ?? lambda.Runtime.PYTHON_3_7;
    const entryFilename = path.basename(props.entry, '.py');

    super(scope, id, {
      ...props,
      runtime,
      code: bundle({
        ...props,
        entry,
        runtime,
      }),
      handler: `${entryFilename}.${handler}`,
    });
  }
}

/**
 * Resolves the given entry file.
 */
function resolveEntry(entry: string): string {
  if (!/\.py$/.test(entry)) {
    throw new Error('Only Python (.py) entry files are supported.');
  }
  if (!fs.existsSync(entry)) {
    throw new Error(`Cannot find entry file at ${entry}`);
  }
  return path.resolve(entry);
}
