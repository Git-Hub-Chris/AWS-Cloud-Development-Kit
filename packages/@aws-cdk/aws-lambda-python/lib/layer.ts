import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import * as cdk from '@aws-cdk/core';
import { bundleDependenciesLayer, hasDependencies, bundlePythonCodeLayer } from './bundling';

/**
 * Properties for PythonDependenciesLayer
 */
export interface PythonLayerVersionProps extends cdk.CopyOptions {
  /**
   * The path to the root directory of the lambda layer.
   */
  readonly entry: string;

  /**
   * The runtimes compatible with this Layer.
   *
   * @default - All runtimes are supported.
   */
  readonly compatibleRuntimes?: lambda.Runtime[];

  /**
   * The description the this Lambda Layer.
   *
   * @default - No description.
   */
  readonly description?: string;

  /**
   * The SPDX licence identifier or URL to the license file for this layer.
   *
   * @default - No license information will be recorded.
   */
  readonly license?: string;

  /**
   * The name of the layer.
   *
   * @default - A name will be generated.
   */
  readonly layerVersionName?: string;

  /**
   * Specify how the layer should bundle the files it finds in the `entry`
   * directory.
   *
   * @default FILES
   */
  readonly bundlingStrategy?: BundlingStrategy;
}

/**
 * Specifies the approach that bundling should take with the files
 */
export enum BundlingStrategy {
  /**
   * Bundle the files by treating them as dependencies
   */
  DEPENDENCIES = 'dependencies',

  /**
   * Bundle the files by copying them directly into the layer
   */
  FILES = 'files',
}

/**
 * A lambda layer version.
 *
 * @experimental
 */
export class PythonLayerVersion extends lambda.LayerVersion {
  constructor(scope: cdk.Construct, id: string, props: PythonLayerVersionProps) {
    const compatibleRuntimes = props.compatibleRuntimes ?? [lambda.Runtime.PYTHON_3_7];

    // Ensure that all compatible runtimes are python
    for (const runtime of compatibleRuntimes) {
      if (runtime && runtime.family !== lambda.RuntimeFamily.PYTHON) {
        throw new Error('Only `PYTHON` runtimes are supported.');
      }
    }

    // Entry and defaults
    const entry = path.resolve(props.entry);
    // Pick the first compatibleRuntime or PYTHON_3_7
    const runtime = compatibleRuntimes[0] ?? lambda.Runtime.PYTHON_3_7;
    const bundlingStrategy = props.bundlingStrategy ?? BundlingStrategy.DEPENDENCIES;

    let code: lambda.AssetCode;
    if (bundlingStrategy == BundlingStrategy.DEPENDENCIES) {
      if (!hasDependencies(entry)) {
        throw new Error(`No dependencies were detected in ${entry}`);
      }

      code = bundleDependenciesLayer({
        entry,
        runtime,
        exclude: props.exclude,
      });
    } else {
      code = bundlePythonCodeLayer({
        entry: props.entry,
        exclude: props.exclude,
        follow: props.follow,
      });
    }

    super(scope, id, {
      ...props,
      compatibleRuntimes,
      code,
    });
  }
}
