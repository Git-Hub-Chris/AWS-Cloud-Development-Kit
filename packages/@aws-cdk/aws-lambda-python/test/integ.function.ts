import * as path from 'path';
import { Runtime } from '@aws-cdk/aws-lambda';
import { App, CfnOutput, Stack, StackProps } from '@aws-cdk/core';
import { Construct } from 'constructs';
import * as lambda from '../lib';

/*
 * Stack verification steps:
 * * aws lambda invoke --function-name <deployed fn name> --invocation-type Event --payload '"OK"' response.json
 */

class TestStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const python36Function = new lambda.PythonFunction(this, 'my_handler', {
      entry: path.join(__dirname, 'lambda-handler'),
      runtime: Runtime.PYTHON_3_6,
    });
    new CfnOutput(this, 'Python36FunctionName', {
      value: python36Function.functionName,
    });

    const python38Function = new lambda.PythonFunction(this, 'my_handler_python_38', {
      entry: path.join(__dirname, 'lambda-handler'),
      runtime: Runtime.PYTHON_3_8,
    });
    new CfnOutput(this, 'Python38FunctionName', {
      value: python38Function.functionName,
    });
  }
}

const app = new App();
new TestStack(app, 'cdk-integ-lambda-python');
app.synth();
