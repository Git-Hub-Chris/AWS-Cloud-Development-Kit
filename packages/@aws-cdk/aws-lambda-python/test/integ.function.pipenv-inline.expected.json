{
  "Resources": {
    "myhandlerServiceRole77891068": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "myhandlerD202FA8E": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters6327fef54763f920a63cab392d9bd4bc9b14917985a3f95700fcd35610b5e83cS3Bucket4A9639D9"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters6327fef54763f920a63cab392d9bd4bc9b14917985a3f95700fcd35610b5e83cS3VersionKeyCEC48E41"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters6327fef54763f920a63cab392d9bd4bc9b14917985a3f95700fcd35610b5e83cS3VersionKeyCEC48E41"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "myhandlerServiceRole77891068",
            "Arn"
          ]
        },
        "Runtime": "python3.7"
      },
      "DependsOn": [
        "myhandlerServiceRole77891068"
      ]
    }
  },
  "Parameters": {
    "AssetParameters6327fef54763f920a63cab392d9bd4bc9b14917985a3f95700fcd35610b5e83cS3Bucket4A9639D9": {
      "Type": "String",
      "Description": "S3 bucket for asset \"6327fef54763f920a63cab392d9bd4bc9b14917985a3f95700fcd35610b5e83c\""
    },
    "AssetParameters6327fef54763f920a63cab392d9bd4bc9b14917985a3f95700fcd35610b5e83cS3VersionKeyCEC48E41": {
      "Type": "String",
      "Description": "S3 key for asset version \"6327fef54763f920a63cab392d9bd4bc9b14917985a3f95700fcd35610b5e83c\""
    },
    "AssetParameters6327fef54763f920a63cab392d9bd4bc9b14917985a3f95700fcd35610b5e83cArtifactHash2886A553": {
      "Type": "String",
      "Description": "Artifact hash for asset \"6327fef54763f920a63cab392d9bd4bc9b14917985a3f95700fcd35610b5e83c\""
    }
  },
  "Outputs": {
    "FunctionName": {
      "Value": {
        "Ref": "myhandlerD202FA8E"
      }
    }
  }
}